### PR TITLE
provider/aws: Add `aws_elasticache_replication_groups` resource

### PR DESCRIPTION
**Pull request by stack72**
**NOTE**: This is the first use case for Replication Groups. This will allow us to specify a set of characteristics for a replication group and a number of nodes in the cluster and Terraform will go and create them. 

This doesn't include:

* creating a replication group from an existing cluster
* adding a new cluster to a replication group

These use cases will follow very shortly!

---------------------

New PR to rework the ElastiCache Replication Groups work. This means we are now starting to create a common schema between ElastiCache Cluster and Replication Groups

The ElastiCache Cluster tests work as expected still after the Schema has been refactored:

```
% make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSElasticacheCluster_'           ✹
==> Checking that code complies with gofmt requirements...
/Users/stacko/Code/go/bin/stringer
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/08/17 15:20:57 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSElasticacheCluster_ -timeout 120m
=== RUN   TestAccAWSElasticacheCluster_basic
--- PASS: TestAccAWSElasticacheCluster_basic (395.41s)
=== RUN   TestAccAWSElasticacheCluster_snapshotsWithUpdates
--- PASS: TestAccAWSElasticacheCluster_snapshotsWithUpdates (812.77s)
=== RUN   TestAccAWSElasticacheCluster_decreasingCacheNodes
--- PASS: TestAccAWSElasticacheCluster_decreasingCacheNodes (860.08s)
=== RUN   TestAccAWSElasticacheCluster_vpc
--- PASS: TestAccAWSElasticacheCluster_vpc (509.98s)
=== RUN   TestAccAWSElasticacheCluster_multiAZInVpc
--- PASS: TestAccAWSElasticacheCluster_multiAZInVpc (879.67s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	3457.932s
```

* [x] Schema (refactor common schema)
* [x] Resource Specific schema pieces
* [x] CRUD
* [x] Acceptance tests
* [x] Documentation

Final test run for the Replication Groups

```
% make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSElasticacheReplicationGroup_'  ✹
==> Checking that code complies with gofmt requirements...
/Users/stacko/Code/go/bin/stringer
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/08/18 13:16:33 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSElasticacheReplicationGroup_ -timeout 120m
=== RUN   TestAccAWSElasticacheReplicationGroup_basic
--- PASS: TestAccAWSElasticacheReplicationGroup_basic (771.49s)
=== RUN   TestAccAWSElasticacheReplicationGroup_updateDescription
--- PASS: TestAccAWSElasticacheReplicationGroup_updateDescription (892.87s)
=== RUN   TestAccAWSElasticacheReplicationGroup_updateNodeSize
--- PASS: TestAccAWSElasticacheReplicationGroup_updateNodeSize (1541.97s)
=== RUN   TestAccAWSElasticacheReplicationGroup_vpc
--- PASS: TestAccAWSElasticacheReplicationGroup_vpc (756.61s)
=== RUN   TestAccAWSElasticacheReplicationGroup_multiAzInVpc
--- PASS: TestAccAWSElasticacheReplicationGroup_multiAzInVpc (1114.94s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	5077.898s
```

### CODE REVIEW

File: builtin/providers/aws/resource_aws_elasticache_replication_group.go [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+
+	resourceSchema["replication_group_description"] = &schema.Schema{
+		Type:     schema.TypeString,
+		Required: true,
+	}
+
+	resourceSchema["number_cache_clusters"] = &schema.Schema{
+		Type:     schema.TypeInt,
+		Required: true,
+		ForceNew: true,
```
Comment (elblivion) [inline builtin/providers/aws/resource_aws_elasticache_replication_group.go]: :-/

You can add read replicas by calling CreateCacheCluster, maybe something for later on: https://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/Replication.AddReadReplica.html#Replication.AddReadReplica.API

File: builtin/providers/aws/resource_aws_elasticache_cluster.go [hunk 2]
```diff
@@ -283,9 +297,9 @@ func resourceAwsElasticacheClusterCreate(d *schema.ResourceData, meta interface{
 		Pending:    pending,
 		Target:     []string{"available"},
 		Refresh:    cacheClusterStateRefreshFunc(conn, d.Id(), "available", pending),
-		Timeout:    10 * time.Minute,
-		Delay:      10 * time.Second,
-		MinTimeout: 3 * time.Second,
+		Timeout:    40 * time.Minute,
+		MinTimeout: 10 * time.Second,
+		Delay:      30 * time.Second,
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_elasticache_cluster.go]: @stack72 are these number changes based on observed slowness while developing?

File: builtin/providers/aws/resource_aws_elasticache_cluster.go [hunk 3]
```diff
@@ -459,9 +482,9 @@ func resourceAwsElasticacheClusterUpdate(d *schema.ResourceData, meta interface{
 			Pending:    pending,
 			Target:     []string{"available"},
 			Refresh:    cacheClusterStateRefreshFunc(conn, d.Id(), "available", pending),
-			Timeout:    5 * time.Minute,
-			Delay:      5 * time.Second,
-			MinTimeout: 3 * time.Second,
+			Timeout:    80 * time.Minute,
+			MinTimeout: 10 * time.Second,
+			Delay:      30 * time.Second,
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_elasticache_cluster.go]: Same Q as above - updates can take twice as long as creates? 🙊

File: builtin/providers/aws/resource_aws_elasticache_cluster.go (hunk 2 above)
Comment (stack72) [inline builtin/providers/aws/resource_aws_elasticache_cluster.go]: yessir!

File: builtin/providers/aws/resource_aws_elasticache_cluster.go (hunk 3 above)
Comment (stack72) [inline builtin/providers/aws/resource_aws_elasticache_cluster.go]: Create:

=== RUN   TestAccAWSElasticacheReplicationGroup_basic
--- PASS: TestAccAWSElasticacheReplicationGroup_basic (771.49s)
=== RUN   TestAccAWSElasticacheReplicationGroup_updateDescription
--- PASS: TestAccAWSElasticacheReplicationGroup_updateDescription (892.87s)

Update: 
=== RUN   TestAccAWSElasticacheReplicationGroup_updateNodeSize
--- PASS: TestAccAWSElasticacheReplicationGroup_updateNodeSize (1541.97s)

File: builtin/providers/aws/resource_aws_elasticache_replication_group.go [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+		Schema: resourceSchema,
+	}
+}
+
+func resourceAwsElasticacheReplicationGroupCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).elasticacheconn
+
+	if d.Get("engine").(string) != "redis" {
+		return fmt.Errorf("The only acceptable Engine type when using Replication Groups is Redis")
+	}
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_elasticache_replication_group.go]: Should be able to add this as a ValidateFunc for plan time validation:

```go
resourceSchema["engine"].ValidateFunc =  // ...
```

File: website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    replication_group_description = "test description"
+    node_type = "cache.m1.small"
+    number_cache_clusters = 2
+    engine = "redis"
+    port = 6379
+    parameter_group_name = "default.redis2.8"
+    availability_zones = ["us-west-2a","us-west-2b"]
+    automatic_failover_enabled = true
+}
+```
```
Comment (phinze) [inline website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown]: Nit: `hclfmt` the example

**Comment by phinze** [comment]
Other than the few super-minor nits inline this LGTM!

File: builtin/providers/aws/resource_aws_elasticache_cluster.go (hunk 3 above)
Comment (phinze) [inline builtin/providers/aws/resource_aws_elasticache_cluster.go]: 🙈

File: website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `replication_group_id` – (Required) The replication group identifier. This parameter is stored as a lowercase string.
+* `replication_group_description` – (Required) A user-created description for the replication group.
+* `number_cache_clusters` - (Required) The number of cache clusters this replication group will initially have.
+ If Multi-AZ is enabled , the value of this parameter must be at least 2. Changing this number will force a new resource
```
Comment (catsby) [inline website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown]: > this replication group will ~~initially~~ have ... Changing this number will force a new resource

Let's remove initially here, unless there is some supported way of expanding it

File: website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+* `node_type` - (Required) The compute and memory capacity of the nodes in the node group.
+* `engine` - (Required) The name of the cache engine to be used for the cache clusters in this replication group. The only valid value is Redis.
```
Comment (catsby) [inline website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown]: Should we even have this then, if the only value accepted is `redis`? Seems like needless work for our users

File: website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+* `automatic_failover_enabled` - (Optional) Specifies whether a read-only replica will be automatically promoted to read/write primary if the existing primary fails. Defaults to `false`.
+* `availability_zones` - (Optional) A list of EC2 availability zones in which the replication group's cache clusters will be created. The order of the availability zones in the list is not important.
+* `engine_version` - (Optional) The version number of the cache engine to be used for the cache clusters in this replication group.
+* `parameter_group_name` - (Optional) The name of the parameter group to associate with this replication group. If this argument is omitted, the default cache parameter group for the specified engine is used.
+* `subnet_group_name` - (Optional) The name of the cache subnet group to be used for the replication group.
+* `security_group_names` - (Optional) A list of cache security group names to associate with this replication group.
+* `security_group_ids` - (Optional) One or more Amazon VPC security groups associated with this replication group.
```
Comment (catsby) [inline website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown]: Any opposition to calling this `vpc_security_group_ids` to be inline with `aws_instance`? Can an EC2 Classic user use `security_group_ids`?

File: website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+* `node_type` - (Required) The compute and memory capacity of the nodes in the node group.
+* `engine` - (Required) The name of the cache engine to be used for the cache clusters in this replication group. The only valid value is Redis.
+* `automatic_failover_enabled` - (Optional) Specifies whether a read-only replica will be automatically promoted to read/write primary if the existing primary fails. Defaults to `false`.
+* `availability_zones` - (Optional) A list of EC2 availability zones in which the replication group's cache clusters will be created. The order of the availability zones in the list is not important.
+* `engine_version` - (Optional) The version number of the cache engine to be used for the cache clusters in this replication group.
+* `parameter_group_name` - (Optional) The name of the parameter group to associate with this replication group. If this argument is omitted, the default cache parameter group for the specified engine is used.
+* `subnet_group_name` - (Optional) The name of the cache subnet group to be used for the replication group.
+* `security_group_names` - (Optional) A list of cache security group names to associate with this replication group.
```
Comment (catsby) [inline website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown]: I assume a VPC user cannot use `security_group_names`, correct? If so we should document that

File: website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown (hunk 6 above)
Comment (stack72) [inline website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown]: Totally agree on this!

File: website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown (hunk 7 above)
Comment (stack72) [inline website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown]: Agreed - removed

File: website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown (hunk 8 above)
Comment (stack72) [inline website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown]: This comes from the common schema - we call it `security_group_ids` in ElastiCache Cluster as well

File: website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown (hunk 9 above)
Comment (stack72) [inline website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown]: Added a note that this should only be used when creating a Replication Group within a VPC

**Comment by brmcconn** [comment]
Any ETA on when this might be available in an upcoming release? I use Terraform for a lot of other AWS resources and it works great. Would love to be able to add ElastiCache to the list, but can't without being able to run replication groups/Redis.

**Comment by stack72** [comment]
Hi @brmcconn 

This has already been released and is available in the 0.7.2 release

P.

**Comment by brmcconn** [comment]
Great, thanks! @stack72. I didn't realize that it was in that release. Just an fyi -  the documentation on the website for elasticache hasn't been updated to include the new funcitonality - https://www.terraform.io/docs/providers/aws/index.html#, but I see the documentation posted here which should be plenty to get me up and running!

**Comment by stack72** [comment]
Hi @brmcconn 

The documentation is linked to on the side bar https://www.terraform.io/docs/providers/aws/r/elasticache_replication_group.html - hope this helps

P.

**Comment by brmcconn** [comment]
Thanks for the link @stack72. It all works great. However, that's not showing up on the website for me just going to www.terraform.io > docs > providers > AWS (the URL is https://www.terraform.io/docs/providers/aws/index.html). You can see the missing link for aws_elasticache_replication_group in my screenshot below:
image

**Comment by eyalzek** [comment]
@brmcconn try clearing your browser cache

**Comment by brmcconn** [comment]
@eyalzek , I gave that a try, but no luck. However, if I go to the website on my phone I see the documentation there, so we'll just call this a client issue on my end.

**Pull request merged**